src/resources/resource_storage.rs

```diff
@@ -10,14 +10,124 @@ use thiserror::Error;
 use super::{PermissionMask, Resource, ResourceType};
 
 /// Unified resource storage for both redirects and scriptlets.
-#[derive(Default)]
+///
+/// By default, this uses an in-memory storage implementation, however this can be changed using
+/// a custom [ResourceStorageBackend] if desired.
 pub struct ResourceStorage {
+    #[cfg(not(feature = "single-thread"))]
+    backend: Box<dyn ResourceStorageBackend + Sync + Send>,
+    #[cfg(feature = "single-thread")]
+    backend: Box<dyn ResourceStorageBackend>,
+}
+
+/// Loads an empty `InMemoryResourceStorage` backend.
+impl Default for ResourceStorage {
+    fn default() -> Self {
+        Self {
+            backend: Box::new(InMemoryResourceStorage::default()),
+        }
+    }
+}
+
+impl ResourceStorage {
+    #[cfg(not(feature = "single-thread"))]
+    pub fn from_backend<S: ResourceStorageBackend + 'static + Sync + Send>(backend: S) -> Self {
+        Self {
+            backend: Box::new(backend),
+        }
+    }
+
+    #[cfg(feature = "single-thread")]
+    pub fn from_backend<S: ResourceStorageBackend + 'static>(backend: S) -> Self {
+        Self {
+            backend: Box::new(backend),
+        }
+    }
+
+    /// Constructor using an `InMemoryResourceStorage` as the backend with the given resources.
+    #[cfg(test)]
+    pub fn in_memory_from_resources(resources: impl IntoIterator<Item = Resource>) -> Self {
+        Self::from_backend(InMemoryResourceStorage::from_resources(resources))
+    }
+}
+
+/// Customizable backend for [Resource] storage.
+/// Custom implementations could be used to enable (for example) sharing of resources between
+/// multiple [crate::Engine]s, an on-disk backend, or special caching behavior.
+pub trait ResourceStorageBackend {
```

Comment:
That trait looks internal to me. Can we don't make it public? I would say the best option is just to provide a few public methods to build `ResourceStorage` without exposing the other struct/traits